### PR TITLE
[MOE Quantization] Update transformers version to 4.40.0

### PR DESCRIPTION
**Pull request by dbogunowicz**
We need to update the transformers version to support QWEN2-MOE model, see: https://github.com/huggingface/transformers/releases/tag/v4.40.0
_(it also, fits into our goal to be constantly matching the latest release)_

#### By default, untie the vocab embedding weights
transformers `4.40.X` has now difficulty saving post one-shot models. This applies to models that are 1) quantized using new (not vLLM) QuantizationModifiers 2) in the "fakequant" state. 
HF developers made changes to the internal logic of parsing "tied weights" (such as embedding and lm_head modules) on `save_pretrained`. Those decisions are buried quite deep in the transformer's codebase. My solution is to untie the weights for the one-shot models on initialization. Benefits: It requires minimal changes on our side, and does not influence the size of saved models on disk (safe-tensors untie weights as well). Downsides: The one-shot process might be slightly less performant because of doubling the memory required to store the embedding layer in CUDA.

### CODE REVIEW

**Review by mgoin — APPROVED** [review]
LGTM once the tests pass!
Export one seems simple
```
ERROR tests/sparseml/export/transformers/test_generative_transformers.py::TestEndToEndExport::test_export_happy_path[roneneldan/TinyStories-1M-text-generation] - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-runner/pytest-0/test_export_happy_path_ronenel0/model/.huggingface/.gitignore.lock'
```

**Review by bfineran — APPROVED** [review]

**Comment by dbogunowicz** [comment]
@mgoin the failure in export-tests looks transient, just fyi

**Review by rahul-tuli — DISMISSED** [review]

**Review by mgoin — APPROVED** [review]

File: src/sparseml/transformers/finetune/text_generation.py
Comment (bfineran) [inline src/sparseml/transformers/finetune/text_generation.py:151]: per @Satrat let's add a warning that this is disabled and can be turned on if using the new modifier

**Review by bfineran — APPROVED** [review]

File: tests/sparseml/transformers/compression/test_quantization.py
Comment (rahul-tuli) [inline tests/sparseml/transformers/compression/test_quantization.py:171]: Could we remove the commented code before merging, ideally git history should be enough to go back and add this in for larger tests?

File: tests/sparseml/transformers/compression/test_quantization.py
Comment (rahul-tuli) [inline tests/sparseml/transformers/compression/test_quantization.py:165]: Not a fan of TODO's in actual code, would be nice if we could remove this and track this internally(maybe via ASANA?)

File: src/sparseml/transformers/finetune/text_generation.py
Comment (rahul-tuli) [reply src/sparseml/transformers/finetune/text_generation.py:151]: Is this argument always `False`? Wouldn't it be nice to somehow allow users to override this? or am I missing something here? 
On a side note it would be nice to have a short comment explaining why we need this to be `False` the PR description covers this but I'd like to keep the reason somewhat close to the code? thoughts?

File: tests/sparseml/transformers/sparsification/test_compress_tensor_utils.py
Comment (rahul-tuli) [inline tests/sparseml/transformers/sparsification/test_compress_tensor_utils.py:80]: Should we really be checking global sparsity here? or does it make more sense to check tensor sparsity for one of the targeted layers, and assert that the tensor_sparsity `isclose` to the actual sparsity specified in the recipe? To me this seems like a random number that could change if we change the model or make any updates like your diff does?

**Review by rahul-tuli — CHANGES_REQUESTED** [review]
Nice! LGTM pending a response/resolution for comments

**Review by Satrat — DISMISSED** [review]

File: tests/sparseml/transformers/sparsification/test_compress_tensor_utils.py
Comment (mgoin) [reply tests/sparseml/transformers/sparsification/test_compress_tensor_utils.py:80]: +1!

**Review by bfineran — DISMISSED** [review]

File: tests/sparseml/transformers/sparsification/test_compress_tensor_utils.py
Comment (rahul-tuli) [inline tests/sparseml/transformers/sparsification/test_compress_tensor_utils.py:82]: We have a function for this, it's called `tensor_sparsity` here https://github.com/neuralmagic/sparseml/blob/c234638a1d7bd276c18bdc46005434fca88f9881/src/sparseml/pytorch/utils/helpers.py#L556

**Review by Satrat — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**
